File: packages/core/src/options.ts
/** Options used when reading values. */
export interface ReadOptions {
  /** Return expired values instead of filtering them out. */
  readonly allowExpired?: boolean;
}

/** Options used when writing values. */
export interface WriteOptions {
  /** Time-to-live in milliseconds. */
  readonly ttlMs?: number;
  /** Replace an existing value when true. */
  readonly overwrite?: boolean;
}

/** Options used when deleting values. */
export interface DeleteOptions {
  /** Throw when the target value does not exist. */
  readonly mustExist?: boolean;
}

/** Options used when scanning by key prefix. */
export interface ScanOptions {
  /** Key prefix to scan. */
  readonly prefix?: string;
  /** Maximum number of results to return. */
  readonly limit?: number;
  /** Include expired values in scan results. */
  readonly includeExpired?: boolean;
}

/** Applies default read options. */
export function normaliseReadOptions(options: ReadOptions = {}): Required<ReadOptions> {
  return {
    allowExpired: options.allowExpired ?? false,
  };
}

/** Applies default write options. */
export function normaliseWriteOptions(options: WriteOptions = {}): WriteOptions {
  const normalised: WriteOptions = {
    overwrite: options.overwrite ?? true,
  };

  if (options.ttlMs !== undefined) {
    return { ...normalised, ttlMs: options.ttlMs };
  }

  return normalised;
}

/** Applies default delete options. */
export function normaliseDeleteOptions(options: DeleteOptions = {}): Required<DeleteOptions> {
  return {
    mustExist: options.mustExist ?? false,
  };
}

/** Applies default scan options. */
export function normaliseScanOptions(options: ScanOptions = {}): ScanOptions {
  const normalised: ScanOptions = {
    includeExpired: options.includeExpired ?? false,
  };

  return {
    ...normalised,
    ...(options.prefix !== undefined ? { prefix: options.prefix } : {}),
    ...(options.limit !== undefined ? { limit: options.limit } : {}),
  };
}

File: crates/yugendb-core/src/options.rs
//! Operation option types.

use std::time::Duration;

/// Options for read operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ReadOptions {
    /// When true, a driver may return expired values.
    pub allow_expired: bool,
}

impl Default for ReadOptions {
    fn default() -> Self {
        Self {
            allow_expired: false,
        }
    }
}

/// Options for write operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct WriteOptions {
    /// Optional time-to-live duration.
    pub ttl: Option<Duration>,
    /// When false, setting an existing key should fail with a conflict.
    pub overwrite: bool,
}

impl Default for WriteOptions {
    fn default() -> Self {
        Self {
            ttl: None,
            overwrite: true,
        }
    }
}

/// Options for delete operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DeleteOptions {
    /// When true, deleting a missing key should fail.
    pub must_exist: bool,
}

impl Default for DeleteOptions {
    fn default() -> Self {
        Self { must_exist: false }
    }
}

/// Options for prefix scans.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ScanOptions {
    /// Prefix to match against keys.
    pub prefix: Option<String>,
    /// Maximum number of records to return.
    pub limit: Option<usize>,
    /// When true, a driver may include expired values.
    pub include_expired: bool,
}
